feat: FIR-43724 implement streaming in node sdk

bogdantruta-firebolt opened the pull request.

stepansergeevitch commented on test/unit/v2/connection.test.ts:
```diff
@@ -445,4 +447,226 @@ describe("Connection V2", () => {
+        ]
+      }),
+      JSON.stringify({
+        message_type: "FINISH_WITH_ERROR"
```
let's add an actual errors field here. And then make sure we report a structured error with all the info

stepansergeevitch reviewed (commented):
Let's add more integration tests
1. Let's query all data types and make sure they are all parsed and converted correctly. See https://github.com/firebolt-db/firebolt-go-sdk/blob/main/fixtures/all_types_query.sql for V2 example and https://github.com/firebolt-db/firebolt-go-sdk/blob/main/fixtures/all_types_query_v0.sql for V1 example of a query
2. Have a syntax error and check that streaming handles it correctly.

stepansergeevitch commented on src/statement/stream/serverSideStream.ts:
```diff
@@ -0,0 +1,65 @@
+                }
+              }
+            } else if (parsed.message_type === "START") {
+              const metaDataJson = JSONbig.parse(line);
```
nit: No need to parse it twice, we already have the result in `parsed`

bogdantruta-firebolt commented:
> Let's add more integration tests
> 
> 1. Let's query all data types and make sure they are all parsed and converted correctly. See https://github.com/firebolt-db/firebolt-go-sdk/blob/main/fixtures/all_types_query.sql for V2 example and https://github.com/firebolt-db/firebolt-go-sdk/blob/main/fixtures/all_types_query_v0.sql for V1 example of a query
> 2. Have a syntax error and check that streaming handles it correctly.

I have added syntax error and streaming error integration tests to this PR, while I created another task to add type fetching tests since there are some issues

ptiurin commented on README.md:
```diff
@@ -617,37 +617,55 @@ await database.delete();
 ### Streaming results
 
-The recommended way to consume query results is by using streams.
+Streaming can only be used with a v2 connection and only using the execute stream method.
```
```suggestion
Streaming can only be used with a v2 connection and only using the `executeStream` method.
```

ptiurin commented on README.md:
```diff
@@ -617,37 +617,55 @@ await database.delete();
 
-The recommended way to consume query results is by using streams.
+Streaming can only be used with a v2 connection and only using the execute stream method.
+The streamResult method from normal the normal execute method returns an in-memory stream of result, rather than dynamically fetching them from the response.
```
Something went wrong here, let's reread this part.

bogdantruta-firebolt replied:
I have redone this part, you can check again

ptiurin commented on README.md:
```diff
@@ -679,22 +697,45 @@ const serializedStream = new SerializeRowStream()
+<a id="in-memory-stream"></a>
+### In-memory stream
+
+When using the streamResult method on the object returned from a simple execute method, the driver will return an in-memory stream of the result. This is useful for small result sets, but not recommended for large result sets.
```
Let's add a line explaining what in-memory stream is. Something like "In this case the whole result will be first fetched in memory and then made available via `streamResult`. This is done for compatibility reasons and has no performance benefits compared to using `fetchResult`".
This way we're answering some questions that we've already had from the customers.

stepansergeevitch reviewed (commented):
One thing I see missing is support for `numeric` as an alias for `Decimal`. We need to add it so that `numeric` values are converted into BigNumber in the same way `Decimal` values are. We also need to extend support for `bigint` and `double precision` type aliases.

goprean commented on README.md:
```diff
@@ -617,37 +617,55 @@ await database.delete();
 ### Streaming results
 
-The recommended way to consume query results is by using streams.
+Streaming can only be used with a v2 connection and only using the `executeStream` method.
```
if this is for developers I don't know if they know what v2 connection means. Should we say: Firebolt 2.0 backend instead?

ptiurin replied:
Old users will know this, but not someone recently joining Firebolt. I think this is fine to assume only new users will read this docstring and omit mentioning v2 altogether. Firebolt 1.0 users are aware new features are rolled out only for firebolt 2.0 now.

stepansergeevitch replied:
Let's remove the mention of v2 completely here. v2 is the only available version right now, v1 is outdated and only exists for customers who were originally on it.
We need to make sure that we raise an exception if a user tries to executeStream with a v1 connection, which explicitly says that streaming is not supported for v1

goprean commented on test/integration/v2/fetchTypes.test.ts:
```diff
@@ -158,4 +159,153 @@ describe("test type casting on fetch", () => {
+      { name: "col_boolean_null", type: "boolean null" },
+      { name: "col_array", type: "array(int)" },
+      // { name: "col_array_null", type: "array(int) null" },
+      // { name: "col_decimal", type: "decimal(38, 30)" },
```
does this need to be commented out?

bogdantruta-firebolt replied:
these are test cases that should work but right now don't. They will be fixed by the ticket referenced above the test

goprean reviewed (approved):
lgtm (although I am novice in node :-) )

stepansergeevitch reviewed (approved):
LGTM. Please fix the comment by @goprean
